src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,375 @@
+    assertEqual "Read the VersionHistoryTable Table" res2 ans2
+    createVersionedTablesTable c
+    nonVersionInsert c
+        (VersionedTables (TableName "user1") (BlockVersion (BlockHeight 0) (ReorgVersion 0)))
+    res3 <- qry_ c "SELECT * FROM VersionedTables;" [RText, RInt, RInt]
+    assertEqual "Read the VersionedTables Table" res3 ans3
+    return $ Right ()
+  where
+    ans1 =
+      [ [ SInt 0
+        , SBlob
+            "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL"
+        ]
+      ]
+    ans2 = [[SInt 0,SInt 0]]
+    ans3 = [[SText "user1",SInt 0,SInt 0]]
+
+
+createBlockHistoryTable :: Database -> IO ()
+createBlockHistoryTable c =
+  exec_
+    c
+    "CREATE TABLE BlockHistory(blockheight UNSIGNED BIGINT PRIMARY KEY,hash BLOB);"
+
+createVersionHistoryTable  :: Database -> IO ()
+createVersionHistoryTable c =
+  exec_
+    c
+    "CREATE TABLE VersionHistory (version UNSIGNED BIGINT PRIMARY KEY,blockheight UNSIGNED BIGINT);"
+
+createVersionedTablesTable  :: Database -> IO ()
+createVersionedTablesTable c =
+  exec_
+    c
+    "CREATE TABLE VersionedTables (tableid TEXT PRIMARY KEY\
+  \ , blockheight UNSIGNED BIGINT\
+  \ , version UNSIGNED BIGINT);"
+
+{-
+
+for ints, investigate which int?
+
+  history of BH pairs  -- BlockHistory
+
+  schema BlockHistory
+
+  create table BlockHistory
+    (block  INT PRIMARY KEY
+    ,**header    varchar<256>) ** investigate best type for hashes
+    (BLOB seems to be the answer.)
+
+
+  history of BV pairs -- VersionHistory
+
+  schema VersionHistory
+
+  create table VersionHistory
+    (version  INT PRIMARY KEY
+    ,block    INT)
+
+  table management table -- VersionedTables
+
+  schema for VersionedTables
+
+  create table VersionedTables
+        (tableid   varchar<256> PRIMARY KEY
+        , block    INT
+        ,version   INT
+        , touchedBlock INT ** maybe
+        , touchedVersion INT ** maybe
+        )
+
+  tableid (primary key)
+
+  touched table
+
+
+
+-}
+
+detectVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockHash
+  -> IO (Either String OnVersionChange)
+detectVersionChange c bRestore hsh  = do
+  let sqlBtoBH [[SInt bh]] = BlockHeight (fromIntegral bh)
+      sqlBtoBH _ = error "sqlBtoBH: wat"
+      sqlVtoV [[SInt version]] = ReorgVersion version
+      sqlVtoV _ = error "sqlVtoV: wat"
+  bCurrent <- sqlBtoBH <$>
+    qry_ c "SELECT max(blockheight) AS current_block_height FROM BlockHistory;" [RInt]
+  vCurrent <- sqlVtoV <$>
+    qry_ c "SELECT max(version) AS current_version FROM VersionHistory;" [RInt]
+
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <-
+    qry c "SELECT COUNT(*)\
+           \ FROM BlockHistory\
+           \ WHERE blockheight = (?)\
+           \ AND hash = (?);" [SInt $ fromIntegral $ pred bRestore, SBlob (encode hsh)] [RInt]
+
+  if historyInvariant /= [[SInt 1]]
```

Comment:
You're probably going to want a function like `expectSing` from pact: https://github.com/kadena-io/pact/blob/fa27df6c325da037c4fae0ff2217912ac58c1fa5/src-ghc/Pact/Persist/SQLite.hs#L113